### PR TITLE
BUG: Resample.aggregate raising TypeError instead of SpecificationError with missing keys dtypes

### LINKED ISSUE #39025: BUG: Missing keys using aggregation dictionary that are unsortable raise TypeError instead of  SpecificationError
- [ ] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
>>> import numpy as np
>>> import pandas as pd
>>>
>>> pd.__version__
'1.3.0.dev0+358.g68db2d26dd'
>>>
>>> df = pd.DataFrame(
...     np.random.randn(1000, 3),
...     index=pd.date_range("1/1/2012", freq="S", periods=1000),
...     columns=[1, "foo", None],
... )
>>> r = df.resample("3T")
>>> r.agg({1: "mean", "foo": "sum"})
                            1        foo
2012-01-01 00:00:00 -0.070340   9.142790
2012-01-01 00:03:00 -0.048538  31.041777
2012-01-01 00:06:00 -0.058046  -0.394660
2012-01-01 00:09:00  0.014268  -8.947485
2012-01-01 00:12:00  0.040080  -6.869409
2012-01-01 00:15:00  0.020587   0.225159
>>>
>>> r.agg({2: "mean", "bar": "sum"})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas\pandas\core\resample.py", line 298, in aggregate
    result, how = aggregate(self, func, *args, **kwargs)
  File "C:\Users\simon\pandas\pandas\core\aggregation.py", line 566, in aggregate
    return agg_dict_like(obj, arg, _axis), True
  File "C:\Users\simon\pandas\pandas\core\aggregation.py", line 741, in agg_dict_like
    cols = sorted(set(keys) - set(selected_obj.columns.intersection(keys)))
TypeError: '<' not supported between instances of 'str' and 'int'
>>>
>>> df = pd.DataFrame(
...     np.random.randn(1000, 3),
...     index=pd.date_range("1/1/2012", freq="S", periods=1000),
...     columns=["A", "B", "C"],
... )
>>>
>>> r = df.resample("3T")
>>> r.agg({"r1": "mean", "r2": "sum"})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas\pandas\core\resample.py", line 298, in aggregate
    result, how = aggregate(self, func, *args, **kwargs)
  File "C:\Users\simon\pandas\pandas\core\aggregation.py", line 566, in aggregate
    return agg_dict_like(obj, arg, _axis), True
  File "C:\Users\simon\pandas\pandas\core\aggregation.py", line 742, in agg_dict_like
    raise SpecificationError(f"Column(s) {cols} do not exist")
pandas.core.base.SpecificationError: Column(s) ['r1', 'r2'] do not exist
>>>
```

code sample based on test_agg_consistency in pandas/tests/resample/test_resample_api.py

#### Problem description

with mypy 0.790

pandas/core/aggregation.py:741: error: Value of type variable "_LT" of "sorted" cannot be "Optional[Hashable]"  [type-var]

#### Expected Output

pandas.core.base.SpecificationError: Column(s) [2, 'bar'] do not exist

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here leaving a blank line after the details tag]

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #39025
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/aggregation.py [hunk 1]
```diff
@@ -738,7 +738,10 @@ def agg_dict_like(
         if isinstance(selected_obj, ABCDataFrame) and len(
             selected_obj.columns.intersection(keys)
         ) != len(keys):
-            cols = sorted(set(keys) - set(selected_obj.columns.intersection(keys)))
+            cols = sorted(
```
Comment (jreback) [inline pandas/core/aggregation.py:741]: can you just use safe_sort?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/aggregation.py (hunk 1 above)
Comment (phofl) [reply pandas/core/aggregation.py]: Yes, of course. Thx. Forgot about safe sort...

**Comment by simonjayhawkins** [comment]
Thanks @phofl for the PR. There is a very similar issue `pandas/core/aggregation.py:485: error: Value of type variable "_LT" of "sorted" cannot be "Optional[Hashable]"  [type-var]` that I didn't yet add an issue for.

There are many tests that go through that path that could be used as the basis for a test for a fix.

<details>

```
=============================================== short test summary info ================================================
FAILED pandas/tests/base/test_misc.py::test_memory_usage[series-with-empty-index] - assert 0 == 108
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 0-ops0-names0] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 0-ops1-names1] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 0-ops2-names2] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 0-ops3-names3] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 1-ops0-names0] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 1-ops1-names1] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 1-ops2-names2] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 1-ops3-names3] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'index'-ops0-names0] - Assertio...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'index'-ops1-names1] - Assertio...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'index'-ops2-names2] - Assertio...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'index'-ops3-names3] - Assertio...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'columns'-ops0-names0] - Assert...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'columns'-ops1-names1] - Assert...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'columns'-ops2-names2] - Assert...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_listlike[axis 'columns'-ops3-names3] - Assert...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_empty_listlike[DataFrame-ops0] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_empty_listlike[DataFrame-ops1] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 0-dict] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 0-Series] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 1-dict] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 1-Series] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 'index'-dict] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 'index'-Series] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 'columns'-dict] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_dictlike[axis 'columns'-Series] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_empty_dictlike[DataFrame-ops1] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_empty_dictlike[DataFrame-ops2] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_empty_dictlike[DataFrame-ops3] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_empty_dictlike[DataFrame-ops4] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_empty_dictlike[DataFrame-ops5] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_and_agg_err[axis 0] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_and_agg_err[axis 1] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_and_agg_err[axis 'index'] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_and_agg_err[axis 'columns'] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_agg_dict_nested_renaming_depr - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[sum-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[max-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[min-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[mean-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[prod-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[std-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[var-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[median-DataFrame] - AssertionE...
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[kurt-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[skew-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[all-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_reducer_raises[any-DataFrame] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-cummax] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-cummin] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-cumprod] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-cumsum] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-diff] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-pct_change] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-rank] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_bad_dtype[DataFrame-<lambda>] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_partial_failure[cummax] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_partial_failure[cummin] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_partial_failure[cumprod] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_partial_failure[cumsum] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_partial_failure[diff] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_partial_failure[pct_change] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_partial_failure[rank] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_missing_columns[axis 0] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_missing_columns[axis 1] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_missing_columns[axis 'index'] - AssertionError
FAILED pandas/tests/frame/apply/test_frame_transform.py::test_transform_missing_columns[axis 'columns'] - AssertionError
FAILED pandas/tests/series/apply/test_series_transform.py::test_transform_none_to_type - AssertionError
```

AssertionError due to `assert False` added to code

</details>

so could either fix here, or shall I create a dedicated bug report?

File: pandas/core/aggregation.py (hunk 1 above)
Comment (phofl) [reply pandas/core/aggregation.py]: Hm had to add list calls

**Comment by phofl** [comment]
I could look into this in the evening, when this pr is not merged until then we could do this here

**Comment by phofl** [comment]
@simonjayhawkins  Fixed it.

I am not quite sure why the test was where it was, so I moved it into the frame folder.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
i know we marked the issues as a regression but that was from 1.0.x, so 1.3 is fine.

**Comment by phofl** [comment]
Thought the same, since only the message is the problem

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**